File: 202. Happy Number/py/isHappy.py
'''
LeetCode problem 202. Happy Number
https://leetcode.com/problems/happy-number/description/
'''


class Solution:
    def isHappy(self, n: int) -> bool:
        results = set()
        while True:
            n = self.squareDigits(n)
            if n == 1:
                return True
            if n in results:
                return False
            results.add(n)

    def squareDigits(self, n: int) -> int:
        return sum(int(c) ** 2 for c in str(n))

File: 202. Happy Number/cs/IsHappy/IsHappy.cs
/*
LeetCode problem 202. Happy Number
https://leetcode.com/problems/happy-number/description/
*/

namespace LeetCode.IsHappy;

public class Solution
{
    public bool IsHappy(int n)
    {
        var results = new HashSet<int>();
        while (true)
        {
            n = n.ToString().Sum(c => (c - '0') * (c - '0'));
            if (n == 1)
                return true;
            if (results.Contains(n))
                return false;
            results.Add(n);
        }
    }
}